botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import logging
+import re
+from enum import Enum
+from functools import lru_cache
+from string import Formatter
+from typing import NamedTuple
+
+from botocore import xform_name
+from botocore.compat import quote, urlparse
+from botocore.exceptions import (
+    EndpointInputParametersError,
+    EndpointResolutionError,
+    RuleSetDefinitionError,
+)
+from botocore.utils import (
+    ArnParser,
+    InvalidArnException,
+    is_valid_ipv4_endpoint_url,
+    is_valid_ipv6_endpoint_url,
+    normalize_url_path,
+    percent_encode,
+)
+
+logger = logging.getLogger(__name__)
+
+TEMPLATE_STRING_RE = re.compile(r"\{[a-zA-Z#]+\}")
+VALID_HOST_LABEL_RE = re.compile(
+    r"^((?!-)[a-zA-Z\d-]{1,63}(?<!-))*((?!-)[a-zA-Z\d-]{1,63}(?<!-))$",
+)
+GET_ATTR_RE = re.compile(r"(\w+)\[(\d+)\]")
+CACHE_SIZE = 100
+
+
+class BaseRule:
+    """A rule within a rule set. All rules contain a conditions property,
+    which can be empty, a partitions JSON and documentation about the rule.
+    """
+
+    def __init__(self, conditions, partitions, documentation=None):
+
+        self.conditions = conditions
+        self.partitions = partitions
+        self.documentation = documentation
+
+    def evaluate(self, scope_vars):
+        raise NotImplementedError()
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object.
+
+        :returns: A bool indicating if the input is a function object.
+        """
+        return isinstance(argument, dict) and "fn" in argument
+
+    def is_ref(self, argument):
```

Comment:
All of these methods operate autonomously from the class. If we move them outside of the class, they'll only need to exist once in the application scope rather than on every instance.

Reply:
Jonas brought up we may be able to look at caching on that as well.

Reply:
Yeah that's fine. I understand from a design perspective why these should be static, but I'm not sure I'm clear on the performance benefits?